### PR TITLE
Adding support for GateDefinition and CallDefinedGate

### PR DESCRIPTION
**Pull request by dberthault**
Needs [this PR](https://github.com/HQSquantumsimulations/qoqo/pull/546)

### CODE REVIEW

File: roqoqo-qasm/tests/gate_defs.qasm [hunk 1]
```diff
@@ -8,6 +8,15 @@ gate ry(theta) a { u3(theta,0,0) a; }
 gate rz(phi) a { u1(phi) a; }
 gate cx c,t { CX c,t; }
 gate h a { u2(0,pi) a; }
+
+gate custom_gate(theta, phi) qb1,qb2
+{
+    h qb1;
+    rx(theta,phi) qb2;
```
Comment (mlodi-hqs) [inline roqoqo-qasm/tests/gate_defs.qasm:15]: rx only takes one input

File: roqoqo-qasm/src/backend.rs [hunk 2]
```diff
@@ -33,6 +35,24 @@ use std::usize;
...
+            already_seen_declarations.push(operation.hqslang().to_string());
+            declarations.push_str(&gate_definition(operation, qasm_version)?);
+            if !declarations.is_empty() {
+                declarations.push('\n');
+            }
+        }
+    }
+    Ok(())
+}
+
```
Comment (mlodi-hqs) [inline roqoqo-qasm/src/backend.rs:55]: I think this has to simply be moved elsewhere in the script, or the docstring won't be related to the Backend struct anymore.

File: roqoqo-qasm/src/parser.rs [hunk 3]
```diff
@@ -162,6 +167,44 @@ fn gate_dispatch(name: &str, params: &[String], qubits: &[usize]) -> Option<Oper
                 CalculatorFloat::ZERO,
             )))
         }
+        "u2" => {
+            let theta = CalculatorFloat::ZERO;
```
Comment (mlodi-hqs) [inline roqoqo-qasm/src/parser.rs:171]: I think theta has to be `pi/2`, but double check the definitions here https://arxiv.org/pdf/1707.03429

File: roqoqo-qasm/src/grammars/qasm2_0.pest [hunk 4]
```diff
@@ -1,25 +1,32 @@
...
+atom               = _{ unary_minus? ~ primary }
+primary            = _{ real | integer | "pi" | "(" ~ expr ~ ")" | unary_op ~ "(" ~ expr ~ ")" }
 
-id = @{ ASCII_ALPHA_LOWER ~ (ASCII_ALPHANUMERIC | "_")* }
-real = @{ "-"? ~ ((ASCII_DIGIT+ ~ "." ~ ASCII_DIGIT*) | (ASCII_DIGIT* ~ "." ~ ASCII_DIGIT+)) ~ (("e" | "E") ~ ("-" | "+")? ~ ASCII_DIGIT+)? }
-integer = @{ ASCII_DIGIT+ }
-WHITESPACE = _{ " " }
-bin_op = _{ "+" | "-" | "*" | "/" | "^" }
+id          = @{ ASCII_ALPHA_LOWER ~ (ASCII_ALPHANUMERIC | "_")* }
+name        = @{ (ASCII_ALPHA_LOWER | ASCII_ALPHA_UPPER) ~ (ASCII_ALPHANUMERIC | "_")* }
```
Comment (mlodi-hqs) [inline roqoqo-qasm/src/grammars/qasm2_0.pest:26]: I checked the grammar and it shouldn't be possible to have an upper case as first letter of the gate name, the definition of `id` in the paper explicitly forbids it. Try removing `name` and replacing it with `id` elsewhere.

**Review by mlodi-hqs — CHANGES_REQUESTED** [review]

File: roqoqo-qasm/src/grammars/qasm2_0.pest (hunk 4 above)
Comment (dberthault) [reply roqoqo-qasm/src/grammars/qasm2_0.pest:26]: What about U and CX ? name is used for the gates that are part of the definition so in 
```
gate u3(theta,phi,lambda) q { U(theta,phi,lambda) q; }
gate u2(phi,lambda) q { U(pi/2,phi,lambda) q; }
gate u1(lambda) q { U(0,0,lambda) q; }
```
name will be used to recognize what is inside the `{}`

Comment (mlodi-hqs) [reply roqoqo-qasm/src/grammars/qasm2_0.pest:26]: Ah right, interesting. Guess we'll keep that then.

**Review by mlodi-hqs — APPROVED** [review]

**Pull request merged**